packages/babel-plugin-proposal-decorators/src/transformer.js

```diff
@@ -1,3 +1,212 @@
+    if (!p.isClassMethod({ kind: "constructor" })) {
+      elements.push(getSingleElementDefinition(p, superRef, fId, file));
+      p.remove();
+    }
+  }
+
+  return t.arrayExpression(elements);
+}
+
+function getConstructorPath(path) {
+  return path
+    .get("body.body")
+    .find(path => path.isClassMethod({ kind: "constructor" }));
+}
+
+const bareSupersVisitor = {
+  CallExpression(path, { initializeInstanceElements }) {
+    if (path.get("callee").isSuper()) {
+      path.insertAfter(t.cloneNode(initializeInstanceElements));
+    }
+  },
+  Function(path) {
+    if (!path.isArrowFunctionExpression()) path.skip();
+  },
+};
+
+function insertInitializeInstanceElements(path, initializeInstanceId) {
+  const isBase = !path.node.superClass;
+  const initializeInstanceElements = t.callExpression(initializeInstanceId, [
+    t.thisExpression(),
+  ]);
+
+  const constructorPath = getConstructorPath(path);
+  if (constructorPath) {
+    if (isBase) {
+      constructorPath
+        .get("body")
+        .unshiftContainer("body", [
+          t.expressionStatement(initializeInstanceElements),
+        ]);
+    } else {
+      constructorPath.traverse(bareSupersVisitor, {
+        initializeInstanceElements,
+      });
+    }
+  } else {
+    const constructor = isBase
+      ? t.classMethod(
+          "constructor",
+          t.identifier("constructor"),
+          [],
+          t.blockStatement([t.expressionStatement(initializeInstanceElements)]),
+        )
+      : t.classMethod(
+          "constructor",
+          t.identifier("constructor"),
+          [t.restElement(t.identifier("args"))],
+          t.blockStatement([
+            t.expressionStatement(
+              t.callExpression(t.Super(), [
+                t.spreadElement(t.identifier("args")),
+              ]),
+            ),
+            t.expressionStatement(initializeInstanceElements),
+          ]),
+        );
+    path.node.body.body.push(constructor);
+  }
+}
+
+function transformClass(path, file) {
+  const isDeclaration = path.node.id && path.isDeclaration();
+  const isStrict = path.isInStrictMode();
+  const { superClass } = path.node;
+
+  path.node.type = "ClassDeclaration";
+  if (!path.node.id) path.node.id = path.scope.generateUidIdentifier("class");
+
+  const initializeId = path.scope.generateUidIdentifier("initialize");
+  const superId =
+    superClass &&
+    path.scope.generateUidIdentifierBasedOnNode(path.node.superClass, "super");
+
+  if (superClass) path.node.superClass = superId;
+
+  const classDecorators = extractDecorators(path);
+  const definitions = getElementsDefinitions(path, path.node.id, file);
+
+  insertInitializeInstanceElements(path, initializeId);
+
+  const expr = template.expression.ast`
+      ${file.addHelper("decorate")}(
+        ${classDecorators || t.nullLiteral()},
+        function (${initializeId}, ${superClass ? superId : null}) {
+          ${isStrict ? null : t.stringLiteral("use strict")}
```

Comment:
This isn't quite the right structure unfortunately, because `use strict` should be a `t.directive`, and also because this will insert it into `body` instead of the `directives` array. I don't think there's an easy way to create an optional directive with babel-templates at the moment unfortunately. I think we'd have to inject it manually after the template AST is created.